lib/Containers/AtomicList.h

```diff
@@ -0,0 +1,279 @@
+        _memoryUsage.fetch_sub(newUsage);
+        auto toDelete = std::move(_history[_ringBufferPos]);
+        _history[_ringBufferPos] = std::move(current);
+        _ringBufferPos = (_ringBufferPos + 1) % _maxHistory;
+        if (toDelete != nullptr) {
+          _trash.emplace_back(std::move(toDelete));
+        }
+        // If somebody prepended something to his copy of _current before
+        // we are completely done here, it will be prepended to the list,
+        // which is already or will be soon in _history.
+      }
+      // Some comments about correctness are in order here: Only one thread
+      // will be in this critical section here at a time. Here, we do the
+      // following operations:
+      //  - copy _current to the ring buffer
+      //  - free a potentially overwritten entry in the ring buffer
+      //  - set _current to a new empty AtomicList
+      // It can happen that another thread copies the shared_ptr in _current
+      // to prepend to the AtomicList, whilst this thread here copies the
+      // shared_ptr to the ring buffer. This is OK.
+      // That is, it is possible that another thread prepends to an AtomicList
+      // which is already in the ring buffer. We need to prove that it cannot
+      // happen that this other thread prepends to the AtomicList and this
+      // thread here destroys it at the same time. But this is guaranteed since
+      // we make a copy of the shared_ptr for each `prepend` call. Everybody
+      // will eventually destroy its shared_ptr and the atomic logic in
+      // shared_ptr will see to it that only one thread frees it. In most
+      // cases this will be the one in the critical section, but even if the
+      // critical section frees its shared_ptr first, then it won't destruct
+      // the AtomicList and the other thread which prepends to it will do it.
+      // This is very unlikely to happen, since `prepend` will probably not
+      // lose a lot of time between copying _current and then prepending
+      // to the AtomicList. Usually, the critical section here will delete
+      // a different AtomicList. But this proof shows that even in the unlikely
+      // event that both work on the same AtomicList, nothing is leaked or
+      // crashes.
+    }
+  }
+
+  // Get current list snapshot
+  std::shared_ptr<AtomicList<T>> getCurrentSnapshot() const {
+    // Use the `getSnapshot` method of the AtomicList and then follow
+    // the Nodes, but do not delete any Node! The allocation is safe as
+    // long as you keep the shared_ptr. Note however, that it is possible
+    // that Nodes are prepended to the list whilst you hold it!
+    // This can throw in out of memory situations.
+    return _current.load(std::memory_order_acquire);
+    // This synchronizes with a store when a new list is stored in the
+    // `prepend` method above.
+  }
+
+  // Get historical snapshots
+  std::vector<std::shared_ptr<AtomicList<T>>> getHistoricalSnapshot() const {
```

Comment:
Instead of this function with a rather complicated return type, we could just hand in a lambda that is executed for every entry. This can then be called via
```
_apiCallRecord->forItem([&](ApiCallRecord record) { // do stuff, e.g. add record to a builder }
```
Then you can also get rid of the `AtomicList::getSnapshot` function.

Reply:
Done.